drivers/wifi/nrf700x/src/fw_load.c

```diff
@@ -66,28 +70,197 @@ LOG_MODULE_DECLARE(wifi_nrf, CONFIG_WIFI_NRF700X_LOG_LEVEL);
+		addr_info = &nrf70_fw_addr_info[1];
+		break;
+	case NRF70_IMAGE_UMAC_PRI:
+		addr_info = &nrf70_fw_addr_info[2];
+		break;
+	case NRF70_IMAGE_UMAC_SEC:
+		addr_info = &nrf70_fw_addr_info[3];
+		break;
+	default:
+		LOG_ERR("Invalid image id: %d\n", image_id);
+		goto out;
+	}
+
+	fw_chunk_info.dest_addr = addr_info->dest_addr + rpu_addr_offset;
+	memcpy(fw_chunk_info.id_str, addr_info->name, sizeof(addr_info->name));
+	fw_chunk_info.data = fw_chunk;
+	fw_chunk_info.size = len;
+
+	status = nrf_wifi_fmac_fw_chunk_load(fmac_dev_ctx,
+					     addr_info->rpu_proc,
+					     &fw_chunk_info);
+	if (status != NRF_WIFI_STATUS_SUCCESS) {
+		LOG_ERR("Failed to load patch chunk, %s", addr_info->name);
+	}
+
+out:
+	return status;
+}
+
+enum nrf_wifi_status nrf_wifi_fw_load(void *rpu_ctx)
+{
+	enum nrf_wifi_status status = NRF_WIFI_STATUS_FAIL;
+	struct nrf70_fw_image_info patch_hdr;
+	unsigned int image_id;
+	int err;
+	unsigned int offset = 0, rpu_addr_offset;
+	struct nrf_wifi_fmac_dev_ctx *fmac_dev_ctx = rpu_ctx;
+	unsigned int max_chunk_size = CONFIG_NRF_WIFI_FW_FLASH_CHUNK_SIZE;
+	char *fw_chunk = NULL;
+
+	err = flash_area_open(NRF70_FW_PATCH_ID, &fa);
+	if (err < 0) {
+		LOG_ERR("Failed to open flash area: %d", err);
+		goto out;
+	}
+
+	LOG_DBG("Flash area opened with size: %d, offset: %ld", fa->fa_size, fa->fa_off);
+
+	/* Read the Header from Flash */
+	err = flash_area_read(fa, 0, &patch_hdr, sizeof(patch_hdr));
+	if (err < 0) {
+		LOG_ERR("Failed to read patch header from flash: %d", err);
+		goto out;
+	}
+	offset += sizeof(patch_hdr);
+
+	status = nrf_wifi_validate_fw_header(rpu_ctx, &patch_hdr);
+	if (status != NRF_WIFI_STATUS_SUCCESS) {
+		LOG_ERR("Failed to validate patch header: %d", status);
+		goto out;
+	}
+
+	status = nrf_wifi_fmac_fw_reset(rpu_ctx);
+	if (status != NRF_WIFI_STATUS_SUCCESS) {
+		LOG_ERR("Failed to reset FMAC: %d", status);
+		goto out;
+	}
+
+	fw_chunk = k_malloc(max_chunk_size);
+	if (!fw_chunk) {
+		LOG_ERR("Failed to allocate memory for patch chunk size: %d", max_chunk_size);
+		goto out;
+	}
+
+	for (image_id = 0; image_id < patch_hdr.num_images; image_id++) {
+		struct nrf70_fw_image image;
+		unsigned int num_chunks;
+		unsigned int chunk_id;
+
+		rpu_addr_offset = 0;
+
+		/* Read sub-header */
+		err = flash_area_read(fa, offset, &image, sizeof(image));
+		if (err < 0) {
+			LOG_ERR("Failed to read patch image from flash: %d", err);
+			goto out;
+		}
+		offset += sizeof(image);
+
+		num_chunks = image.len / max_chunk_size +
+			     (image.len % max_chunk_size ? 1 : 0);
+		LOG_DBG("Processing image %d, len: %d, num_chunks: %d",
+			image_id, image.len, num_chunks);
+		for (chunk_id = 0; chunk_id < num_chunks; chunk_id++) {
+			unsigned int chunk_size = image.len - chunk_id * max_chunk_size;
+
+			if (chunk_size > max_chunk_size) {
+				chunk_size = max_chunk_size;
+			}
+
+			LOG_DBG("Processing chunk %d-%d, size: %d", image_id, chunk_id, chunk_size);
+			memset(fw_chunk, 0, max_chunk_size);
```

Comment:
I guess that you are filling the buffer with 0 to have 0 for everything that has not been filled with data.
Assuming that most read operations will fill the buffer with data completely, it may be more efficient to call the memset post `nrf_wifi_read_and_download_chunk` to fill what has left of the buffer.

Reply:
TBH, we don't need this at all, as we are passing the length, so for the last chunk (all others chunks will have full size) the rest of the bytes will be ignored. I will just remove it.